src/spartan/batched.rs

```diff
@@ -0,0 +1,645 @@
+
+    // Create evaluation instances for W(r_y[1..]) and E(r_x)
+    let (w_vec, u_vec) = {
+      let mut w_vec = Vec::with_capacity(2 * num_instances);
+      let mut u_vec = Vec::with_capacity(2 * num_instances);
+      w_vec.extend(polys_W.into_iter().map(|poly| PolyEvalWitness { p: poly }));
+      u_vec.extend(
+        evals_W
+          .iter()
+          .zip(U.iter())
+          .zip(r_y)
+          .map(|((&eval, u), r_y)| PolyEvalInstance {
+            c: u.comm_W,
+            x: r_y[1..].to_vec(),
+            e: eval,
+          }),
+      );
+
+      w_vec.extend(polys_E.into_iter().map(|poly| PolyEvalWitness { p: poly }));
+      u_vec.extend(
+        evals_E
+          .iter()
+          .zip(U.iter())
+          .zip(r_x)
+          .map(|((eval_E, u), r_x)| PolyEvalInstance {
+            c: u.comm_E,
+            x: r_x,
+            e: *eval_E,
+          }),
+      );
+      (w_vec, u_vec)
+    };
+
+    let (batched_u, batched_w, sc_proof_batch, claims_batch_left) =
+      batch_eval_prove(u_vec, w_vec, &mut transcript)?;
+
+    let eval_arg = EE::prove(
+      ck,
+      &pk.pk_ee,
+      &mut transcript,
+      &batched_u.c,
+      &batched_w.p,
+      &batched_u.x,
+      &batched_u.e,
+    )?;
+
+    let (evals_Az, evals_Bz, evals_Cz): (Vec<_>, Vec<_>, Vec<_>) =
+      evals_Az_Bz_Cz.into_iter().multiunzip();
+
+    Ok(BatchedRelaxedR1CSSNARK {
+      sc_proof_outer,
+      claims_outer: (evals_Az, evals_Bz, evals_Cz),
+      evals_E,
+      sc_proof_inner,
+      evals_W,
+      sc_proof_batch,
+      evals_batch: claims_batch_left,
+      eval_arg,
+    })
+  }
+
+  fn verify(&self, vk: &Self::VerifierKey, U: &[RelaxedR1CSInstance<G>]) -> Result<(), NovaError> {
+    let mut transcript = G::TE::new(b"BatchedRelaxedR1CSSNARK");
+
+    transcript.absorb(b"vk", &vk.digest());
+    U.iter().for_each(|u| {
+      transcript.absorb(b"U", u);
+    });
+
+    let num_instances = U.len();
+
+    let (num_rounds_x, num_rounds_y): (Vec<_>, Vec<_>) = vk
+      .S
+      .iter()
+      .map(|s| (s.num_cons.log_2(), s.num_vars.log_2() + 1))
+      .unzip();
+    let num_rounds_x_max = *num_rounds_x.iter().max().unwrap();
+    let num_rounds_y_max = *num_rounds_y.iter().max().unwrap();
+
+    // Define τ polynomials of the appropriate size for each instance
+    let polys_tau = {
+      let tau = transcript.squeeze(b"t")?;
+
+      num_rounds_x
+        .iter()
+        .map(|&num_rounds| PowPolynomial::new(&tau, num_rounds))
+        .collect::<Vec<_>>()
+    };
+
+    // Sample challenge for random linear-combination of outer claims
+    let outer_r = transcript.squeeze(b"out_r")?;
+    let outer_r_powers = powers::<G>(&outer_r, num_instances);
+
+    let (claim_outer_final, r_x) = self.sc_proof_outer.verify_batch(
+      &vec![G::Scalar::ZERO; num_instances],
+      &num_rounds_x,
+      &outer_r_powers,
+      3,
+      &mut transcript,
+    )?;
+
+    // Collect evaluation point for each instance
+    let r_x = num_rounds_x
```

Comment:
A little more documentation of exactly what's going on here would not hurt.

Reply:
https://hackmd.io/@adr1anh/BJw1g0aBT#Non-uniform-instances